### ISSUE TITLE
Invalid instantiation of generic indexed access type gives "any" without warning

### ISSUE THREAD

**Issue by mattmccutchen** [Microsoft/TypeScript#25673]
**TypeScript Version:**  master (60986ad)

**Search Terms:** invalid instantiation generic index indexed access lookup type intersection any warning error

**Code** (compile with `--noImplicitAny`)
```ts
function index<T extends {[P in K]: {}}, K extends string>(t: T, k: K): T[K] {
    return t[k];
}
function index2<T, K extends keyof T>(t: T, k: K): T[K] {
    return t[k];
}
let t = {a: 42};
let k: "a" & {} = undefined;
let x0 = t[k];          // x0 is any, error TS7017
let x1 = index(t, k);   // x1 is any, no error
let x2 = index2(t, k);  // x2 is any, no error
```

**Expected behavior:** Either x1 and x2 produce an error/warning or their types are something other than `any`.

**Actual behavior:** As marked above.

**Playground Link:** [Link](https://www.typescriptlang.org/play/#src=function%20index%3CT%20extends%20%7B%5BP%20in%20K%5D%3A%20%7B%7D%7D%2C%20K%20extends%20string%3E(t%3A%20T%2C%20k%3A%20K)%3A%20T%5BK%5D%20%7B%0D%0A%20%20%20%20return%20t%5Bk%5D%3B%0D%0A%7D%0D%0Afunction%20index2%3CT%2C%20K%20extends%20keyof%20T%3E(t%3A%20T%2C%20k%3A%20K)%3A%20T%5BK%5D%20%7B%0D%0A%20%20%20%20return%20t%5Bk%5D%3B%0D%0A%7D%0D%0Alet%20t%20%3D%20%7Ba%3A%2042%7D%3B%0D%0Alet%20k%3A%20%22a%22%20%26%20%7B%7D%20%3D%20undefined%3B%0D%0Alet%20x0%20%3D%20t%5Bk%5D%3B%20%20%20%20%20%20%20%20%20%20%2F%2F%20x0%20is%20any%2C%20error%20TS7017%0D%0Alet%20x1%20%3D%20index(t%2C%20k)%3B%20%20%20%2F%2F%20x1%20is%20any%2C%20no%20error%0D%0Alet%20x2%20%3D%20index2(t%2C%20k)%3B%20%20%2F%2F%20x2%20is%20any%2C%20no%20error) (remember to enable `--noImplicitAny`)

**Related Issues:** Didn't find any.

**Comment by sandersn** [comment]
Here's a modified repro that (1) gets rid of the mapped type constraint (2) gets rid of the incorrect initialiser for `k`. I think the problem is that indexing returns errorType and the caller doesn't log an error when that happens.

```ts
function index<T extends {[s: string]: {}}, K extends string>(t: T, k: K): T[K] {
    return t[k];
}
function index2<T, K extends keyof T>(t: T, k: K): T[K] {
    return t[k];
}
let t = {a: 42};
declare let k: "a" & {};
let x0 = t[k];          // x0 is any, error TS7017
let x1 = index(t, k);   // x1 is any, no error
let x2 = index2(t, k);  // x2 is any, no error
```

**Comment by sandersn** [comment]
Well, this is an instance of the general problem we have with reporting errors on type instantiation. In general, it can occur as a result of chain or tree of instantiations, far away from either the usage or the declaration. In this case, however, we are close enough to the instantiation to confidently say that `index` is the best error location.

I'm going to try threading an error location  through `checkCallExpression >>= getReturnTypeOfSignature >>= instantiateType >>= instantiateTypeWorker >>= getIndexedAccessType >>= getPropertyTypeForIndexType` [1]. It will probably fix the bug, and may be useful enough to ship. However, I don't think it will scale well: if we start passing in lots of error locations, the chance that we'll start reporting errors in bizarre locations goes up. For example, if we were to add an error location to instantiation of conditional types, I'm certain we would see examples of an error happening on a call expression that should actually be on the original type alias, or perhaps even a derived type alias.

[1] Using a parameter, not the State monad. :frowning_woman:

**Comment by sandersn** [comment]
Well, I talked about this idea with @RyanCavanaugh and we think it might work out. I put what I have in the branch `sandersn/better-instantiateType-errors` and I'll come back to it after I have a chance to finish up a couple other things.

**Comment by sandersn** [comment]
Narrator: It did not work out.

**Comment by sandersn** [comment]
However, these examples do now produce `unknown` at least, which is an improvement.
